### PR TITLE
All prep required for individual comments

### PR DESCRIPTION
**Pull request by stkent**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -4,7 +4,7 @@ A Gradle plugin that helps facilitate Github PR checking and automatic commentin
 ## Usage
 
 Gnag is meant to be simple to use and easy to drop in to any Android project. Shown below is the simplest 
-Gnag setup that will report violations to GitHub. By default this config will report PMD, Findbugs, Checkstyle and 
+Gnag setup that will violations violations to GitHub. By default this config will violations PMD, Findbugs, Checkstyle and
```
Comment (stkent) [inline README.md]: Fix this rename

File: README.md [hunk 2]
```diff
@@ -34,12 +34,12 @@ This is the simplest way to add automatic PR checking and commenting to your pro
 
 #### Tasks
 
-You can use the gnagCheck gradle task to run Gnag locally and generate an HTML report in the build directory. 
+You can use the gnagCheck gradle task to run Gnag locally and generate an HTML violations in the build directory.
```
Comment (stkent) [inline README.md]: Fix this rename

File: README.md [hunk 3]
```diff
@@ -34,12 +34,12 @@ This is the simplest way to add automatic PR checking and commenting to your pro
...
 #### Tasks
 
-You can use the gnagCheck gradle task to run Gnag locally and generate an HTML report in the build directory. 
+You can use the gnagCheck gradle task to run Gnag locally and generate an HTML violations in the build directory. 
 ```groovy
 ./gradlew clean gnagCheck
 ```
 
-You can use the gnagReport task which will first run gnagCheck and then report this to the Github issue specified. 
+You can use the gnagReport task which will first run gnagCheck and then violations this to the Github issue specified.
```
Comment (stkent) [inline README.md]: Fix this rename

File: README.md [hunk 4]
```diff
@@ -76,21 +76,21 @@ gnag {
 ```
 
 - ***enabled*** - easily disable Gnag in specific situations
-- ***failOnError*** - should reporter violations cause the build to fail or just generate a report
-- ***checkstyle*** - block to customize the checkstyle reporter
+- ***failOnError*** - should violationDetector violations cause the build to fail or just generate a violations
```
Comment (stkent) [inline README.md]: Fix this grammar

File: README.md [hunk 5]
```diff
@@ -76,21 +76,21 @@ gnag {
...
   - ***enabled*** - set if findbugs should execute
   - ***reporterConfig*** - provide a custom [findbugs config](http://findbugs.sourceforge.net/manual/filter.html)
-- ***androidLint*** - block to customize the android lint reporter
-  - ***enabled*** - set if the android lint reporter should look for a lint report
+- ***androidLint*** - block to customize the android lint violationDetector
+  - ***enabled*** - set if the android lint violationDetector should look for a lint violations
   - ***severity*** - can be 'Error' or 'Warning' depending on which severity you want Gnag to check
 - ***github*** - block to customize Github reporting (only used during the `gnagReport` task
-  - ***repoName*** - account and repo name to report violations to
+  - ***repoName*** - account and repo name to violations violations to
```
Comment (stkent) [inline README.md]: Fix this rename

File: README.md [hunk 6]
```diff
@@ -76,21 +76,21 @@ gnag {
...
   - ***enabled*** - set if pmd should execute
   - ***reporterConfig*** - provide a custom [pmd config](http://pmd.sourceforge.net/pmd-5.1.1/howtomakearuleset.html)
-- ***findbugs*** - block to customize the findbugs reporter
+- ***findbugs*** - block to customize the findbugs violationDetector
   - ***enabled*** - set if findbugs should execute
   - ***reporterConfig*** - provide a custom [findbugs config](http://findbugs.sourceforge.net/manual/filter.html)
-- ***androidLint*** - block to customize the android lint reporter
-  - ***enabled*** - set if the android lint reporter should look for a lint report
+- ***androidLint*** - block to customize the android lint violationDetector
+  - ***enabled*** - set if the android lint violationDetector should look for a lint violations
```
Comment (stkent) [inline README.md]: Fix this rename, grammar

File: README.md [hunk 7]
```diff
@@ -76,21 +76,21 @@ gnag {
...
+- ***failOnError*** - should violationDetector violations cause the build to fail or just generate a violations
+- ***checkstyle*** - block to customize the checkstyle violationDetector
   - ***enabled*** - set if checkstyle should execute
   - ***reporterConfig*** - provide a custom [checkstyle config](http://checkstyle.sourceforge.net/config.html)
-- ***pmd*** - block to customize the pmd reporter
+- ***pmd*** - block to customize the pmd violationDetector
   - ***enabled*** - set if pmd should execute
   - ***reporterConfig*** - provide a custom [pmd config](http://pmd.sourceforge.net/pmd-5.1.1/howtomakearuleset.html)
-- ***findbugs*** - block to customize the findbugs reporter
+- ***findbugs*** - block to customize the findbugs violationDetector
```
Comment (stkent) [inline README.md]: Fix this grammar

File: README.md [hunk 8]
```diff
@@ -76,21 +76,21 @@ gnag {
...
 
 - ***enabled*** - easily disable Gnag in specific situations
-- ***failOnError*** - should reporter violations cause the build to fail or just generate a report
-- ***checkstyle*** - block to customize the checkstyle reporter
+- ***failOnError*** - should violationDetector violations cause the build to fail or just generate a violations
+- ***checkstyle*** - block to customize the checkstyle violationDetector
   - ***enabled*** - set if checkstyle should execute
   - ***reporterConfig*** - provide a custom [checkstyle config](http://checkstyle.sourceforge.net/config.html)
-- ***pmd*** - block to customize the pmd reporter
+- ***pmd*** - block to customize the pmd violationDetector
```
Comment (stkent) [inline README.md]: Fix this grammar

File: README.md [hunk 9]
```diff
@@ -76,21 +76,21 @@ gnag {
 ```
 
 - ***enabled*** - easily disable Gnag in specific situations
-- ***failOnError*** - should reporter violations cause the build to fail or just generate a report
-- ***checkstyle*** - block to customize the checkstyle reporter
+- ***failOnError*** - should violationDetector violations cause the build to fail or just generate a violations
+- ***checkstyle*** - block to customize the checkstyle violationDetector
```
Comment (stkent) [inline README.md]: Fix this grammar

File: plugin/build.gradle [hunk 10]
```diff
@@ -15,6 +15,7 @@ compileJava {
 }
 
 repositories {
+    mavenLocal()
```
Comment (stkent) [inline plugin/build.gradle]: Remove this before review

File: plugin/build.gradle [hunk 11]
```diff
@@ -29,6 +30,8 @@ dependencies {
 
     compile 'com.squareup.okhttp3:logging-interceptor:3.2.0'
 
+    compile 'com.github.stkent:githubdiffparser:1.2'
```
Comment (stkent) [inline plugin/build.gradle]: Safe to remove before review

File: plugin/src/main/groovy/com/btkelly/gnag/api/GitHubApi.java [hunk 12]
```diff
@@ -95,4 +92,16 @@ public GitHubPullRequest getPullRequestDetails() {
         }
     }
 
+    public GitHubPullRequestDiffWrapper getPullRequestDiffWrapper() {
```
Comment (stkent) [inline plugin/src/main/groovy/com/btkelly/gnag/api/GitHubApi.java]: Currently unused

File: plugin/src/main/groovy/com/btkelly/gnag/api/GitHubApiClient.java [hunk 13]
```diff
@@ -38,4 +39,7 @@
     @GET("pulls/{issueNumber}")
     Call<GitHubPullRequest> getPullRequest(@Path("issueNumber") String issueNumber);
 
+    @GET("pulls/{issueNumber}")
```
Comment (stkent) [inline plugin/src/main/groovy/com/btkelly/gnag/api/GitHubApiClient.java]: Currently unused

File: plugin/src/main/groovy/com/btkelly/gnag/models/CheckStatus.java [hunk 14]
```diff
@@ -15,29 +15,50 @@
...
-    private final String comment;
+    @NotNull
     private final GitHubStatusType gitHubStatusType;
 
-    public CheckStatus(String comment, GitHubStatusType gitHubStatusType) {
-        this.comment = comment;
-        this.gitHubStatusType = gitHubStatusType;
-    }
+    @NotNull
+    private final Set<Violation> violations;
```
Comment (stkent) [inline plugin/src/main/groovy/com/btkelly/gnag/models/CheckStatus.java:38]: Replaces the String comment

File: plugin/src/main/groovy/com/btkelly/gnag/models/CheckStatus.java [hunk 15]
```diff
@@ -15,29 +15,50 @@
...
 
-    public String getComment() {
-        return comment;
+    public CheckStatus(@NotNull final GitHubStatusType gitHubStatusType, @NotNull final Set<Violation> violations) {
+        if (gitHubStatusType == SUCCESS && !violations.isEmpty()) {
+            throw new IllegalStateException("Status cannot be SUCCESS if violations are found.");
```
Comment (stkent) [inline plugin/src/main/groovy/com/btkelly/gnag/models/CheckStatus.java:42]: Consistency check

File: plugin/src/main/groovy/com/btkelly/gnag/models/GitHubPullRequestDiffWrapper.java [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+     *     The position value is the number of lines down from the first "@@" hunk header in the file you would like to
+     *     comment on. The line just below the "@@" line is position 1, the next line is position 2, and so on.
+     *     The position in the file's diff continues to increase through lines of whitespace and additional hunks until
+     *     a new file is reached.
+     *     
+     * 
+     * 
+     */
+
+    public GitHubPullRequestDiffWrapper(
```
Comment (stkent) [inline plugin/src/main/groovy/com/btkelly/gnag/models/GitHubPullRequestDiffWrapper.java]: Currently unused

File: plugin/src/main/groovy/com/btkelly/gnag/tasks/GnagReportTask.java [hunk 17]
```diff
@@ -31,13 +32,15 @@
...
 import static com.btkelly.gnag.models.GitHubStatusType.PENDING;
+import static com.btkelly.gnag.models.GitHubStatusType.SUCCESS;
 
 /**
  * Created by bobbake4 on 4/1/16.
  */
 public class GnagReportTask extends DefaultTask {
 
     public static final String TASK_NAME = "gnagReport";
+    private static final String REMOTE_SUCCESS_COMMENT = "Congrats, no poop code found! This PR is safe to merge.";
```
Comment (stkent) [inline plugin/src/main/groovy/com/btkelly/gnag/tasks/GnagReportTask.java]: Should have :poop:

File: plugin/src/main/groovy/com/btkelly/gnag/tasks/GnagReportTask.java [hunk 18]
```diff
@@ -61,15 +64,22 @@ public void taskAction() {
...
-            CheckStatus checkStatus = (CheckStatus) projectStatus;
-
+            final CheckStatus checkStatus = (CheckStatus) projectStatus;
             System.out.println("Project status: " + checkStatus);
 
-            gitHubApi.postGitHubComment(checkStatus.getComment());
+            if (checkStatus.getGitHubStatusType() == SUCCESS) {
+                gitHubApi.postGitHubComment(REMOTE_SUCCESS_COMMENT);
+            } else {
+                // TODO: send individual message(s) here when possible
```
Comment (stkent) [inline plugin/src/main/groovy/com/btkelly/gnag/tasks/GnagReportTask.java:76]: The magic happens here later

File: plugin/src/main/groovy/com/btkelly/gnag/utils/ReportWriter.java [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Set;
+
+/**
+ * Created by bobbake4 on 4/18/16.
+ */
+public final class ReportWriter {
```
Comment (stkent) [inline plugin/src/main/groovy/com/btkelly/gnag/utils/ReportWriter.java:30]: Used for writing a report to a local dir

File: .idea/codeStyleSettings.xml [hunk 20]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project version="4">
+  <component name="ProjectCodeStyleSettingsManager">
+    <option name="PER_PROJECT_SETTINGS">
+      <value />
+    </option>
+    <option name="PREFERRED_PROJECT_CODE_STYLE" value="Stuart" />
```
Comment (btkelly) [inline .idea/codeStyleSettings.xml:7]: Huh?

Comment (stkent) [inline .idea/codeStyleSettings.xml:7]: :trollface:

File: plugin/src/main/groovy/com/btkelly/gnag/reporters/AndroidLintViolationDetector.groovy [hunk 21]
```diff
@@ -52,15 +52,39 @@ class AndroidLintReporter implements Reporter {
...
+                        try {
+                            lineNumber = violation.location.@line.toInteger()
+                        } catch (final NumberFormatException e) {
+                            System.out.println("Error reading line number from Android Lint violations.");
+                            e.printStackTrace();
+                            lineNumber = null
+                        }
+
+                        result.add(new Violation(
+                                sanitize((String) violation.@id.text()),
```
Comment (btkelly) [inline plugin/src/main/groovy/com/btkelly/gnag/reporters/AndroidLintViolationDetector.groovy:73]: Is there a reason the sanitize function isn't just inside the Violation constructor?

Comment (stkent) [inline plugin/src/main/groovy/com/btkelly/gnag/reporters/AndroidLintViolationDetector.groovy:73]: Thought about it, but it seemed weird to transform automatically. It's also called earlier in the stream for the FindbugsViolationDetector, which needs to do some funky stuff to guess the correct file path.

**Pull request merged**